swarm/src/connection.rs

```diff
@@ -54,28 +56,26 @@ pub struct Connected {
 /// Event generated by a [`Connection`].
 #[derive(Debug, Clone)]
 pub enum Event<T> {
-    /// Event generated by the [`ConnectionHandler`].
+    /// Event generated by the [`NodeHandlerWrapper`].
     Handler(T),
     /// Address of the remote has changed.
     AddressChange(Multiaddr),
 }
 
-/// A multiplexed connection to a peer with an associated `ConnectionHandler`.
-pub struct Connection<TMuxer, THandler>
+/// A multiplexed connection to a peer with an associated [`NodeHandlerWrapper`].
+pub struct Connection<THandler>
 where
-    TMuxer: StreamMuxer,
-    THandler: ConnectionHandler<Substream = Substream<TMuxer>>,
+    THandler: ProtocolsHandler,
 {
     /// Node that handles the muxing.
-    muxing: substream::Muxing<TMuxer, THandler::OutboundOpenInfo>,
+    muxing: substream::Muxing<StreamMuxerBox, NodeHandlerWrapperOutboundOpenInfo<THandler>>,
     /// Handler that processes substreams.
-    handler: THandler,
+    handler: NodeHandlerWrapper<THandler>,
```

Comment:
Could we inline this struct into `Connection` now? Then we could just get rid of `NodeHandlerWrapper` completely!

I think the only reason we are passing this around at the moment is because we are constructing it "too early". We could just pass `THandler` around everywhere have all other fields of `NodeHandlerWrapper` be in `Connection`.

Reply:
> I think the only reason we are passing this around at the moment is because we are constructing it "too early".

I think you are confusing the concept of `IntoProtocolsHandler` -> `ProtocolsHandler`, `NodeHandlerWrapperBuilder` -> `NodeHandlerWrapper` and now obsolete `IntoConnectionHandler` -> `ConnectionHandler` with the relation between `ProtocolsHandler` and `NodeHandlerWrapper`, right?

(Puuuh, so many `XXXHandler`. I hope we end up with only `ProtocolsHandler` (renamed to `ConnectionHandler`) once all the refactorings are in.)

> Could we inline this struct into `Connection` now?

I think so. At least move the two closer together. I suggest we rename `ProtocolsHandler` to `ConnectionHandler` first ( including some smaller clean-ups) and then consider folding `NodeHandlerWrapper` into `Connection`.

Reply:
I may confuse some things yes 😁

Either order is fine with me, it was a more general remark, not suggesting it should be done next or at all! :)